### PR TITLE
Added composite command support and test coverage

### PR DESCRIPTION
**Pull request by msivasubramaniaan**
### What does this PR do?
This PR supports to run the composite commands in terminal

### What issues does this PR fix?
https://github.com/eclipse-che/che/issues/21859
https://github.com/eclipse-che/che/issues/23726
https://github.com/eclipse-che/che/issues/23709
https://github.com/eclipse-che/che/issues/23725

### How to test this PR?
Added [workspace](https://workspaces.openshift.com/#https://github.com/msivasubramaniaan/devfile-test-command/tree/main?editor-image=quay.io/che-incubator-pull-requests/che-code:pr-639-amd64) which has all commands of devfile. Run those commands from `devfile.yaml`.

### Does this PR contain changes that override default upstream Code-OSS behavior?
- [ ] the PR contains changes in the [code](https://github.com/che-incubator/che-code/tree/main/code) folder (you can skip it if your changes are placed in a che extension )
- [ ] the corresponding items were added to the [CHANGELOG.md](https://github.com/che-incubator/che-code/blob/main/.rebase/CHANGELOG.md) file
- [ ] rules for automatic `git rebase` were added to the [.rebase](https://github.com/che-incubator/che-code/tree/main/.rebase) folder

## Summary by CodeRabbit

* **New Features**
  * Composite devfile commands: sequential or parallel execution across components with validation, flattening, cancellation and failure handling

* **Tests**
  * Added comprehensive Jest suites, mocks and TS/Jest configs covering composite execution, task creation, env handling, shell-preservation, failure semantics, naming, nesting and edge cases

* **Chores**
  * Build images now run extension tests during image build; added test script and Jest tooling to the extension package

### CODE REVIEW

File: code/extensions/che-commands/package-lock.json [hunk 1]
```diff
@@ -38,701 +537,3821 @@
...
-      "version": "1.3.0",
-      "resolved": "https://registry.npmjs.org/@jsep-plugin/assignment/-/assignment-1.3.0.tgz",
-      "integrity": "sha512-VVgV+CXrhbMI3aSusQyclHkenWSAm95WaiKrMxRFam3JSUiIaQjoMIw2sEs/OX4XifnqeQUN4DYbJjlA8EfktQ==",
-      "engines": {
-        "node": ">= 10.16.0"
-      },
-      "peerDependencies": {
-        "jsep": "^0.4.0||^1.0.0"
+    "node_modules/@emnapi/core": {
+      "version": "1.8.1",
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/package-lock.json]: FYI
The license check is failing

Image

File: code/extensions/che-commands/src/taskProvider.ts [hunk 2]
```diff
@@ -10,86 +10,422 @@
...
+								'echo "Composite multi-component execution completed"',
+								this.expandEnvVariables("${PROJECT_SOURCE}"),
+							);
+						}),
+					);
 				}
+
+				// Single-component → join commands
+				const joiner = parallel ? " & " : " && ";
+				let compositeCmd = execs.map((e) => e.commandLine).join(joiner);
```
Comment (rgrunber) [inline code/extensions/che-commands/src/taskProvider.ts]: @RomanNikitenko , is string manipulation the only way to do this ? Using compound tasks on the vscode side still isn't possible ?

Comment (RomanNikitenko) [reply code/extensions/che-commands/src/taskProvider.ts]: I investigated that area more than 2 years ago, so I asked @msivasubramaniaan to double check that option when he started to work on the issue.
@msivasubramaniaan could you provide result of your investigation?

Comment (msivasubramaniaan) [reply code/extensions/che-commands/src/taskProvider.ts]: I double-checked this again.

At the moment, yes — string manipulation is still the only practical way to achieve this.

VS Code does not support dynamically creating or chaining compound tasks programmatically from an extension. Compound tasks must be statically defined in tasks.json, and there’s no API to generate or invoke a compound task on the fly based on resolved components or runtime data.

Because of that limitation, joining commands using `&& / &` (depending on parallel vs sequential execution) remains the only viable solution when the task needs to be assembled dynamically.

So the approach of building a composite command string is the only way

**Review by RomanNikitenko — COMMENTED** [review]
@msivasubramaniaan 
I tried to test the functionality using your samples
I run `component-parallel-demo` command:

Image

My understanding - it should run `signal-frontend` in the `frontend` container and `signal-backend` in the `backend` container.

Actual behaviour: 
- there is no output for both commands, for both containers
- there is just message "Composite multi-component execution completed" which comes from the `taskProvider` https://github.com/che-incubator/che-code/pull/639/changes#diff-e505171c2bbd2db1494bbb49e7c7ccaad44644b923a12f6c5ca2ebdedd0fdbcdR389

Moreover, the dependent commands just hanging without any output when I run them separately:

Image

Could you clarify:
- what is expected behaviour there? 
- is it possible to detect which component is used for your test commands?
- If commands are in the running state - why composite command reports that `Composite multi-component execution completed`?

thanks in advance!

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
I've prepared own sample and tested `parallel` and `sequential` uses cases for 2 components, I can confirm - it does not work, there is a bug - no output for both dependent commands and it's unclear if the commands are executed at all:

Screenshot 2026-02-02 at 23 17 53

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
I've detected one more bug : a command `label` should be displayed instead of `id`:

Screenshot 2026-02-02 at 23 29 09
 
 
Screenshot 2026-02-02 at 23 34 23

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
I see you've requested a new review, but image for testing is not ready.
Please let me know when you test all use cases we discussed and regressions are fixed.
Thanks!

**Comment by msivasubramaniaan** [comment]
Hello @RomanNikitenko 

As discussed, the PR is ready for composite commands. I've tested with various cases and confirmed that all cases were working. Here is the proof for `sequential flow with cross component`
image

But still facing the issue with parallel flow of cross component. Due to PTY limitation on Devspaces the output got mixed. 
image
 if needed we can raise separate issue for track

**Review by RomanNikitenko — COMMENTED** [review]
> As discussed, the PR is ready for composite commands. I've tested with various cases and confirmed that all cases were working. Here is the proof for sequential flow with cross component

sorry, but `sequential flow with cross component` does not work correctly

- let's say a user need to build a project first and then run it
- let's say the build process takes some time 
- all this time a user has no output for the build process
- it shows the output when the first task is completed only
- but then there is another bug: a user has no output for the second task
- again - it shows the output when a user stops the second task

Please see a short video.
- I created a simple task that prints logs every second to simulate a long running task (like build)
- First I run that regular task to show how it works
- then - I run a composite task

https://github.com/user-attachments/assets/e0d74ab7-ba16-43d0-aea0-aaaecfee9902

**Comment by msivasubramaniaan** [comment]
> > As discussed, the PR is ready for composite commands. I've tested with various cases and confirmed that all cases were working. Here is the proof for sequential flow with cross component
> 
> sorry, but `sequential flow with cross component` does not work correctly
> 
> * let's say a user need to build a project first and then run it
> * let's say the build process takes some time
> * all this time a user has no output for the build process
> * it shows the output when the first task is completed only
> * but then there is another bug: a user has no output for the second task
> * again - it shows the output when a user stops the second task
> 
> Please see a short video.
> 
> * I created a simple task that prints logs every second to simulate a long running task (like build)
> * First I run that regular task to show how it works
> * then - I run a composite task
> 
>  composite.long.task.bug.mp4

Hello @RomanNikitenko 
I wanted to know that both tasks are using `same component` or `different component`? When I checked with `same component` it is working as expected. Please share the `devfile` if possible.

https://github.com/user-attachments/assets/4e482251-eeb4-45e9-b85b-2c12f744719f

**Comment by RomanNikitenko** [comment]
> Hello @RomanNikitenko I wanted to know that both tasks are using `same component` or `different component`? When I checked with `same component` it is working as expected. Please share the `devfile` if possible.
> 

Hello @msivasubramaniaan!

Let's go through our discussion step by step:
- you wrote "Here is the proof for sequential flow with `cross` component" in the https://github.com/che-incubator/che-code/pull/639#issuecomment-3890384463 
- you added a screenshot with the `Sequential: Check echo ubi8-ubi9` command to confirm - it works for `different` components
- I answered with the message "sorry, but sequential flow with `cross` component does not work correctly"
- I provided a video to report that the functionality does not work for sequential long-running composite commands with `different` components

So - I meant `different` components.
I'm sorry if I misunderstood your initial comment with the screenshot.

**Comment by msivasubramaniaan** [comment]
> > Hello @RomanNikitenko I wanted to know that both tasks are using `same component` or `different component`? When I checked with `same component` it is working as expected. Please share the `devfile` if possible.
> 
> Hello @msivasubramaniaan!
> 
> Let's go through our discussion step by step:
> 
> * you wrote "Here is the proof for sequential flow with `cross` component" in the [Added composite command support and test coverage #639 (comment)](https://github.com/che-incubator/che-code/pull/639#issuecomment-3890384463)
> * you added a screenshot with the `Sequential: Check echo ubi8-ubi9` command to confirm - it works for `different` components
> * I answered with the message "sorry, but sequential flow with `cross` component does not work correctly"
> * I provided a video to report that the functionality does not work for sequential long-running composite commands with `different` components
> 
> So - I meant `different` components. I'm sorry if I misunderstood your initial comment with the screenshot.

@RomanNikitenko **I have fixed the issue. Now user can able to see the logs.**
image

**Review by RomanNikitenko — COMMENTED** [review]
> @RomanNikitenko I have fixed the issue. Now user can able to see the logs.

I tested that use case again and can confirm - it's fixed.
Thank you!

Within testing that use case I detected another problem:
- it's not possible to stop a composite task using Ctrl+C or button on a command item
- processes are hanging for long running commands even after using `Kill` button on UI
- so - no command item on UI, but the corresponding process continue to work/is hanging

Please see this video:

https://github.com/user-attachments/assets/800a9575-5143-42fa-918c-fa38c5ae11c8

**Comment by msivasubramaniaan** [comment]
> > @RomanNikitenko I have fixed the issue. Now user can able to see the logs.
> 
> I tested that use case again and can confirm - it's fixed. Thank you!
> 
> Within testing that use case I detected another problem:
> 
> * it's not possible to stop a composite task using Ctrl+C or button on a command item
> * processes are hanging for long running commands even after using `Kill` button on UI
> * so - no command item on UI, but the corresponding process continue to work/is hanging
> 
> Please see this video:
> 
>  Stop.command.bug.mp4

Hello @RomanNikitenko 
I can handle the both case of terminating the process. Please refer the screenshot:
image

**Comment by msivasubramaniaan** [comment]
Hello @RomanNikitenko 

The parallel on cross component working as expected. 
image

But due to concurrency issue on the shared PTY, I have made when user runs parallel cross component, it opens multiple terminals, basically each one for a command. And we can see the merged output in the parent terminal. You can see it in the attachement.  

https://github.com/user-attachments/assets/b38d22eb-eb00-4bb4-ad51-85ff8d6a9525

**Review by RomanNikitenko — COMMENTED** [review]
I guess the following logs were added just to debug within implementation
please remove them - it's hard to find useful information in the `Devfile Commands` channel due to tons of `debug`-logs - I don't think it's useful for a regular user...

https://github.com/user-attachments/assets/f32bf754-1519-459a-9fd6-8474a17b6b46

**Comment by RomanNikitenko** [comment]
For one composite task there are at least 3 terminals
The first one shows mixed output for all dependent task.

Let's say dependent tasks build different modules:
- do we really need all mixed output in the first terminal?
- how it's going to be used? I guess it's hard to find any useful information because of tons mixed logs...
wdyt?

https://github.com/user-attachments/assets/a762e0a1-9eca-42e5-b6d4-622411ad1fbb

**Comment by RomanNikitenko** [comment]
Regarding dependent commands labels:

Screenshot 2026-02-24 at 12 43 41

- a dependent task item does not look readable - `devfile: parallel-ubi8-ubi9:ubi8-tools(web-nodejs-sample)`
- in the current implementation each command has its own terminal - do we really need that prefix for dependent tasks? (like: `parallel-ubi8-ubi9`- is `id` of a `composite` task - is it required for a `dependent` task item?)
- I propose to use command `label` instead of `id` to get it more readable, like - `devfile: ubi8 container echo` seems clearer than `devfile: parallel-ubi8-ubi9:ubi8-tools(web-nodejs-sample)`

**Comment by RomanNikitenko** [comment]
I think there is inconsistent behavior related to the `scanning task output` prompt. This prompt is not displayed when I run a `simple` devfile task, but it is displayed when I run a `composite` task.

https://github.com/user-attachments/assets/4b4bf54a-88c6-440d-9c79-0e240f6f6337

**Comment by msivasubramaniaan** [comment]
> For one composite task there are at least 3 terminals The first one shows mixed output for all dependent task.
> 
> Let's say dependent tasks build different modules:
> 
> * do we really need all mixed output in the first terminal?
> * how it's going to be used? I guess it's hard to find any useful information because of tons mixed logs...
>   wdyt?
> 
>  mixed_output.mov


For a composite task running dependent tasks in parallel, there are indeed multiple terminals:

- Each dependent task runs in its own terminal (isolated execution).

- The first terminal shows aggregated output from all dependent tasks.

**Why we keep the mixed output in the first terminal:**

- The composite task is treated as a single logical operation, so the main terminal provides a unified view of progress and results.
- Users can monitor the overall execution without switching between multiple terminals.
- It behaves similarly to many build tools (like Maven/Gradle/npm workspaces) where logs from multiple modules appear together.

**How it can be used effectively:**

1. The aggregated output is tagged by component, for example:

```
[backend] build started
[frontend] build started
[backend] compiling...
[frontend] bundling...
[backend] done
[frontend] done

```

2. This makes it possible to identify which component produced each log line.
3. If detailed debugging is needed, users can still open the individual task terminals, which contain the full isolated logs.

**Why this approach is useful:**

- Gives a quick overview of the composite execution.
- Avoids constantly switching terminals.
- Preserves full logs per component when needed.

So the first terminal is intended as a **high-level monitoring** view, while the individual terminals remain available for detailed investigation.

**Comment by RomanNikitenko** [comment]
> **How it can be used effectively:**
> 
> 1. The aggregated output is tagged by component, for example:
> 
> ```
> [backend] build started
> [frontend] build started
> [backend] compiling...
> [frontend] bundling...
> [backend] done
> [frontend] done
> ```
> 
> 2. This makes it possible to identify which component produced each log line.

Sorry - but in the current implementation there is no `aggregated output is tagged by component`
So - it's not true: `This makes it possible to identify which component produced each log line.`

@msivasubramaniaan 
I understand that there is a technical issue and you are trying to workaround it
but
if you would have a choice - would you use the mixed output or separate?

my personal opinion - there are already separate terminals for each dependent task with the corresponding output - I would not use mixed output in the third terminal...

**Comment by msivasubramaniaan** [comment]
> > **How it can be used effectively:**
> > 
> > 1. The aggregated output is tagged by component, for example:
> > 
> > ```
> > [backend] build started
> > [frontend] build started
> > [backend] compiling...
> > [frontend] bundling...
> > [backend] done
> > [frontend] done
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. This makes it possible to identify which component produced each log line.
> 
> Sorry - but in the current implementation there is no `aggregated output is tagged by component` So - it's not true: `This makes it possible to identify which component produced each log line.`
> 
> @msivasubramaniaan I understand that there is a technical issue and you are trying to workaround it but if you would have a choice - would you use the mixed output or separate?
> 
> my personal opinion - there are already separate terminals for each dependent task with the corresponding output - I would not use mixed output in the third terminal...

@RomanNikitenko
Here is the reference of Parallel cross component used only equal no of terminals for the list of commands.
image

Also the inconsistency behavior issue also fixed and removed the unnecessary logs as well

File: code/extensions/che-commands/src/compositeTaskBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+
+				onDidClose: closeEmitter.event,
+
+				open: () => {
+					execs.forEach((e, index) => {
+						const childTask = this.buildExecTask(e, index === 0);
+
+						vscode.tasks.executeTask(childTask);
+					});
+					closeEmitter.fire(0);
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/src/compositeTaskBuilder.ts]: The most critical use case is support for Parallel Composite tasks.
In my view, the current implementation would be the best one:
- it is aligned with the VS Code behavior for compound tasks
- dependent tasks run in their own terminals
- there is no mixed output in a single terminal (as in the previous implementation)
- there are no extra terminals: number of terminals = number of dependent tasks (vs the previous approach, where it was number of dependent tasks  + 1 parent terminal)

Unfortunately, the current approach has one significant drawback: if I understand correctly - actually a parent (extra) terminal is still created and then immediately closed. It looks like a bug on UI.

https://github.com/user-attachments/assets/7008c4d9-7550-4946-a830-c5f13b153435

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
One more use case I detected is:
- let's say there is a classical `build` and `run` sequential command
- current behavior: the second command runs even when the first one fails
- I expect: `run` command doesn't start - it does not make sense to run it when the previous one (`build`) fails
- I checked that behavior for the vanilla VS Code compound task - VS Code doesn't start the second command if the first one fails

Please see on the video 2 behaviors:
- the first one - for the che-code composite task
- second one - for the vanilla VS Code compound task

https://github.com/user-attachments/assets/e3a14ac1-5e5d-4a36-ab33-d4c96afa0464

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
Another problem for the sequential composite command support:
- first command fails but second still runs
- composite item is displayed as successful one
- it's possible to detect that something went wrong only searching the corresponding logs

Screenshot 2026-02-27 at 14 06 55

Update: it does not matter first command fails or the second one - command item has incorrect status...
On the screenshot bellow the second command failed

Screenshot 2026-02-27 at 14 33 45

  

VS Code behavior, for example:

Screenshot 2026-02-27 at 14 10 07

**Comment by msivasubramaniaan** [comment]
Hello @RomanNikitenko
As discussed completed all the feedbacks other than https://github.com/che-incubator/che-code/pull/639#discussion_r2863404377. AFAIK `VS Code requires a task to have an execution to run`, so I just created a task and close it immediately. If any other suggestion I am glad to receive it.

File: code/extensions/che-commands/src/compositeTaskBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+				: (vscode.TaskRevealKind.Silent ?? 2),
+			panel: vscode.TaskPanelKind.Dedicated ?? 1,
+			clear: false,
+		};
+
+		return task;
+	}
+
+	private sanitize(cmd: string) {
+		return cmd.replace(/(?:\s*(?:&&|\|\||[|;&]))+\s*$/, "").trim();
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/src/compositeTaskBuilder.ts]: @msivasubramaniaan 
I guess this logic changes behavior from background/non-blocking to foreground/blocking, which is a semantic regression.
Like: `cmd &` becomes `cmd`

Could you double check it using logs or another way?

I prepared simple commands that show a background command blocks next command

https://github.com/user-attachments/assets/54aebc8f-78a1-4a98-9806-c24479b3e168

Comment (RomanNikitenko) [reply code/extensions/che-commands/src/compositeTaskBuilder.ts]: @msivasubramaniaan 
For example, we use background command to run machine-exec in the che-code: https://github.com/che-incubator/che-code/blob/main/build/scripts/entrypoint-volume.sh#L74

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
I detected another bug, it's related to resolving `workingDir`.

On the video you can see that:
- resolving `workingDir` works correctly when I run separate tasks
- there is a bug when I run the same tasks within a composite task


https://github.com/user-attachments/assets/f35c1ea7-3f9b-4a35-887f-eb109fb9eacc

File: code/extensions/che-commands/src/compositeTaskBuilder.ts (hunk 4 above)
Comment (RomanNikitenko) [reply code/extensions/che-commands/src/compositeTaskBuilder.ts]: I used the following commands for testing:
```
- id: bg-command
    exec:
      label: "Background Command"
      component: ubi9-tools
      commandLine: "sh -c 'echo Start Background Task; sleep 5; echo Complete Background Task' &"
  - id: after-bg-command
    exec:
      label: "NEXT Task after BG Command"
      component: ubi9-tools
      commandLine: "sh -c 'sleep 1; echo NEXT Task after BG TASK'"
  - id: bg-in-composite
    composite:
      label: "Background operator stripped"
      commands:
        - bg-command
        - after-bg-command
      parallel: false
```

**Comment by msivasubramaniaan** [comment]
@RomanNikitenko 
I have addressed the review comments which you provide, Here are the references of the fixes:

image

image

**Review by RomanNikitenko — COMMENTED** [review]
@msivasubramaniaan 
I tested commands with env variables, for example `go` stack uses this approach: https://github.com/devfile/registry/blob/7c358cd8ef31c59111fac73808eaab7b37afd19c/stacks/go/2.6.0/devfile.yaml#L66-L73

It works for regular commands, 
but
env variables are not resolved for composite commands

https://github.com/user-attachments/assets/8da4c325-75fa-4afd-b583-a059304ffafd

**Comment by msivasubramaniaan** [comment]
> @msivasubramaniaan I tested commands with env variables, for example `go` stack uses this approach: https://github.com/devfile/registry/blob/7c358cd8ef31c59111fac73808eaab7b37afd19c/stacks/go/2.6.0/devfile.yaml#L66-L73
> 
> It works for regular commands, but env variables are not resolved for composite commands
> 
>  resolving_env_var_bug.mov

@RomanNikitenko The mentioned `env` issue was fixed.
image

File: code/extensions/che-commands/src/compositeTaskBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+			const pty: vscode.Pseudoterminal = {
+				onDidWrite: new vscode.EventEmitter<string>().event,
+				onDidClose: closeEmitter.event,
+
+				open: () => {
+					execs.forEach((e, index) => {
+						const childTask = this.buildExecTask(
+							{
+								...e,
+								command: this.buildEnvPrefix(e.env) + e.command
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/src/compositeTaskBuilder.ts]: it looks like `this.buildEnvPrefix(e.env)` is executed twice
first time here, 
the second one - https://github.com/msivasubramaniaan/che-code/blob/2d5486b1ccea247ba04e4cdce60bac0ef8ce0296/code/extensions/che-commands/src/compositeTaskBuilder.ts#L271

please double check this logic

File: code/extensions/che-commands/src/compositeTaskBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+			},
+			vscode.TaskScope.Workspace,
+			command.composite.label || command.id,
+			"devfile",
+			execution,
+			[],
+		);
+	}
+	
+	buildEnvPrefix(env: any[] | undefined) {
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/src/compositeTaskBuilder.ts]: minor: private?

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
there is still a problem with resolving env variables for the `workingDir` field (not sure if it's the same problem I reported above or a new one)

I think it should be resolved here: https://github.com/msivasubramaniaan/che-code/blob/2d5486b1ccea247ba04e4cdce60bac0ef8ce0296/code/extensions/che-commands/src/compositeTaskBuilder.ts#L106

See an example here https://github.com/msivasubramaniaan/che-code/blob/2d5486b1ccea247ba04e4cdce60bac0ef8ce0296/code/extensions/che-commands/src/taskProvider.ts#L107

 Steps to reproduce:
- define env var on the `container` level in your devfile
- use the variable in the `workingDir` field of a command
- it works when a user runs it as a regular command
- it does not work when a user runs the same command within a composite command

https://github.com/user-attachments/assets/b4e13edf-8c73-4222-b93f-65dc886ae216

**Comment by msivasubramaniaan** [comment]
> @msivasubramaniaan there is still a problem with resolving env variables for the `workingDir` field (not sure if it's the same problem I reported above or a new one)
> 
> I think it should be resolved here: https://github.com/msivasubramaniaan/che-code/blob/2d5486b1ccea247ba04e4cdce60bac0ef8ce0296/code/extensions/che-commands/src/compositeTaskBuilder.ts#L106
> 
> See an example here https://github.com/msivasubramaniaan/che-code/blob/2d5486b1ccea247ba04e4cdce60bac0ef8ce0296/code/extensions/che-commands/src/taskProvider.ts#L107
> 
> Steps to reproduce:
> 
> * define env var on the `container` level in your devfile
> * use the variable in the `workingDir` field of a command
> * it works when a user runs it as a regular command
> * it does not work when a user runs the same command within a composite command
> 
>  resolve_container_env_var.mov

Hello @RomanNikitenko 
As discussed over our standup, please test and merge this PR till before env variable for the `workingDir`. I will made a fix and create separate PR for that.

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
> As discussed over our standup, please test and merge this PR till before env variable for the workingDir

Apart of ^
there are still unresolved comments above. 

Did you have a chance to debug/investigate:
> this.buildEnvPrefix(e.env) is executed twice

?

**Comment by msivasubramaniaan** [comment]
Hello @RomanNikitenko 
I have addressed all the review comments on this PR and tested the `env variable` on container as well. The screenshot FYR
image

**Review by RomanNikitenko — COMMENTED** [review]
@msivasubramaniaan 
it looks like the last changes bring a regression for regular (not composite) commands

I tested simple command: `commandLine: "echo resolved=${HOME} actual=$HOME"` 
The same variable in the same `commandLine` shows different values
I think it's resolved for a wrong container

Image

File: code/extensions/che-commands/src/taskProvider.ts [hunk 7]
```diff
@@ -35,14 +38,65 @@ export class DevfileTaskProvider implements vscode.TaskProvider {
 	private async computeTasks(): Promise<vscode.Task[]> {
 		const devfileCommands = await this.fetchDevfileCommands();
 
+		const devfileService = this.cheAPI.getDevfileService();
+		const devfile = await devfileService.get();
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/src/taskProvider.ts]: This change introduces _double_ `devfileService.get()` call for _one_ `computeTasks()` call:

computeTasks() called
- fetchDevfileCommands()
    - devfileService.get()          ← first call
- devfileService.get()          ← second call (same data, wasted)

I think it makes sense to refactor the logic here

File: code/extensions/che-commands/tests/compositeTaskBuilder.test.ts [hunk 8]
```diff
@@ -0,0 +1,625 @@
+/**********************************************************************
+ * Copyright (c) 2026 Red Hat, Inc.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/tests/compositeTaskBuilder.test.ts:9]: The copyright header here uses the short form, but the rest of the codebase uses the full EPL-2.0 header. Please update for consistency

File: code/extensions/che-commands/jest.config.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    tsconfig: 'tsconfig.json' }],
+};
+export const moduleNameMapper = {
+    '^vscode$': '<rootDir>/tests/__mocks__/vscode.ts'
+};
+export const collectCoverageFrom = [
+    "src/**/*.ts",
+    "!src/**/*.d.ts",
+    "!tests/**",
+]
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/jest.config.js]: Missing trailing newline at end of file. Most editors and POSIX tools expect one, and git will flag it in diffs.

File: code/extensions/che-commands/package.json [hunk 10]
```diff
@@ -75,4 +79,4 @@
       }
     ]
   }
-}
+}
```
Comment (RomanNikitenko) [inline code/extensions/che-commands/package.json]: Missing trailing newline at end of file.

**Comment by tolusha** [comment]
Hi! I'm **che-ai-assistant** — I help with your pull requests.

**Available commands**:
- `/che-ai-assistant generate-che-doc` — Generate a documentation PR based on this PR's changes
- `/che-ai-assistant ok-pr-review` — Run a comprehensive PR review (summary, code review, deep review, impact analysis)
- `/che-ai-assistant help` — Show this help message

**Comment by msivasubramaniaan** [comment]
Hello @RomanNikitenko,
I have changed the resolver logic to handle the `env variable` when it not available in the context/container level. Here I have attached both screenshots and tested other `env based commands` as well and confirmed there were no breaks/regression.

**Devfile:** 
image
**CMD:** 
image

File: code/extensions/che-commands/tests/compositeTaskBuilder.test.ts (hunk 8 above)
Comment (RomanNikitenko) [reply code/extensions/che-commands/tests/compositeTaskBuilder.test.ts:9]: @msivasubramaniaan 
you resolved the comment, but did not provide the change - it's still actual

**Review by RomanNikitenko — COMMENTED** [review]
> Here I have attached both screenshots and tested other env based commands as well and confirmed there were no breaks/regression.

@msivasubramaniaan 
Sorry, but the screenshots indicate that there are now two issues instead of the single issue I reported above. In both cases, the variables were resolved for the wrong container, so the situation has actually become worse.

- In the `ubi` container, `/opt/app-root/src` is the `HOME` directory.
- In the `alpine` container, `/` is the `HOME` directory.

In your case both variables in the `alpine` container were resolved to the `/opt/app-root/src`...

With this in mind, could you please review the implementation and verify whether the same approach is used elsewhere? We should ensure that the correct container is consistently selected for each step, from variable resolution through command execution.

**Comment by RomanNikitenko** [comment]
I tested use case from PR with fix `Environment variable is not set when running a task`, see https://github.com/che-incubator/che-code/pull/367
- `devfile:test` works well for the `main`
- it fails for the current PR

`main`:
image

`current PR`
image

**Comment by RomanNikitenko** [comment]
Another use case I took also from the https://github.com/che-incubator/che-code/pull/367
- there is `build` command in the devfile
- it works as a simple command (not composite)
- it fails when the same command is included in a composite command

https://github.com/user-attachments/assets/5465f4fb-8c6d-424f-8992-7d39b9cd434b

**Comment by msivasubramaniaan** [comment]
> Another use case I took also from the #367
> 
> * there is `build` command in the devfile
> * it works as a simple command (not composite)
> * it fails when the same command is included in a composite command
> 
>  composite_cmd_fails.mov

The issue was resolved. Please refer the video,

https://github.com/user-attachments/assets/2be3f80f-123c-4740-bdd4-ce5f13bc86c8

**Comment by msivasubramaniaan** [comment]
> > Here I have attached both screenshots and tested other env based commands as well and confirmed there were no breaks/regression.
> 
> @msivasubramaniaan Sorry, but the screenshots indicate that there are now two issues instead of the single issue I reported above. In both cases, the variables were resolved for the wrong container, so the situation has actually become worse.
> 
> * In the `ubi` container, `/opt/app-root/src` is the `HOME` directory.
> * In the `alpine` container, `/` is the `HOME` directory.
> 
> In your case both variables in the `alpine` container were resolved to the `/opt/app-root/src`...
> 
> With this in mind, could you please review the implementation and verify whether the same approach is used elsewhere? We should ensure that the correct container is consistently selected for each step, from variable resolution through command execution.

The issue was resolved. 
image

**Comment by RomanNikitenko** [comment]
@msivasubramaniaan 
I've merged a lot commits from the upstream into the main branch.
Please align your branch with the `main` one.
Also there is a conflict that must be resolved.

I'm going to do another cycle of testing when your branch is updated.
Thanks!
